File: problems-adapt/0601-0700/0640_count-difference-bounded-pairs/solution_merge_counting.go
func countDifferenceBoundedPairs(nums1 []int, nums2 []int, diff int) int64 {
	values := make([]int64, len(nums1))
	for i := range values {
		values[i] = int64(nums1[i]) - int64(nums2[i])
	}
	count := int64(0)

	// mergeSort sorts values[lo:hi) while tallying, into count, the cross pairs
	// values[i] <= values[j] + diff with i in the left half and j in the right.
	var mergeSort func(lo, hi int)
	mergeSort = func(lo, hi int) {
		if hi-lo < 2 {
			return
		}
		mid := (lo + hi) / 2
		mergeSort(lo, mid)
		mergeSort(mid, hi)
		left := make([]int64, mid-lo)
		copy(left, values[lo:mid])
		p := 0 // left values at or below the running bound
		for j := mid; j < hi; j++ {
			for p < len(left) && left[p] <= values[j]+int64(diff) {
				p++
			}
			count += int64(p) // each admitted left value pairs with this right element
		}
		i, j, k := 0, mid, lo
		for i < len(left) && j < hi {
			if left[i] <= values[j] { // equal: the left element places first
				values[k] = left[i]
				i++
			} else {
				values[k] = values[j]
				j++
			}
			k++
		}
		for i < len(left) {
			values[k] = left[i]
			i++
			k++
		}
	}
	mergeSort(0, len(values))
	return count
}

File: problems-adapt/2401-2500/2426_count-difference-bounded-pairs/solution_merge_counting.py
class Solution:
    def countDifferenceBoundedPairs(self, nums1: list[int], nums2: list[int], diff: int) -> int:
        values = [x - y for x, y in zip(nums1, nums2)]

        def merge_sort(lo, hi):
            """Sort values[lo:hi], returning the cross pairs values[i] <= values[j] + diff."""
            if hi - lo < 2:
                return 0
            mid = (lo + hi) // 2
            count = merge_sort(lo, mid) + merge_sort(mid, hi)
            left = values[lo:mid]
            p = 0  # left values at or below the running bound
            for j in range(mid, hi):
                while p < len(left) and left[p] <= values[j] + diff:
                    p += 1
                count += p  # each admitted left value pairs with this right element
            i, j, k = 0, mid, lo
            while i < len(left) and j < hi:
                if left[i] <= values[j]:  # equal: the left element places first
                    values[k] = left[i]
                    i += 1
                else:
                    values[k] = values[j]
                    j += 1
                k += 1
            while i < len(left):
                values[k] = left[i]
                i += 1
                k += 1
            return count

        return merge_sort(0, len(values))
